packages/aws_signature_v4/lib/src/request/canonical_request/canonical_request.dart

```diff
@@ -0,0 +1,268 @@
+    required AWSCredentials credentials,
+    required AWSCredentialScope credentialScope,
+    required AWSAlgorithm algorithm,
+    required Duration expiresIn,
+    int contentLength = 0,
+    String payloadHash = emptyPayloadHash,
+    ServiceConfiguration serviceConfiguration =
+        const BaseServiceConfiguration(),
+  }) {
+    final headers = Map.of(request.headers);
+    final queryParameters = Map.of(request.queryParameters);
+
+    // Per https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html
+    assert(
+      expiresIn > const Duration(seconds: 1) &&
+          expiresIn < const Duration(days: 7),
+      'Expiration must be greater than 1 second and less than 7 days',
+    );
+
+    final canonicalHeaders = CanonicalHeaders(headers);
+    final signedHeaders = SignedHeaders(canonicalHeaders);
+    serviceConfiguration.applyPresigned(
+      queryParameters,
+      request: request,
+      credentialScope: credentialScope,
+      algorithm: algorithm,
+      expiresIn: expiresIn.inSeconds,
+      signedHeaders: signedHeaders,
+      credentials: credentials,
+    );
+
+    return CanonicalRequest._(
+      request: request,
+      queryParameters: queryParameters,
+      canonicalQueryParameters: CanonicalQueryParameters(queryParameters),
+      headers: headers,
+      canonicalHeaders: canonicalHeaders,
+      signedHeaders: signedHeaders,
+      credentialScope: credentialScope,
+      algorithm: algorithm,
+      expiresIn: expiresIn,
+      presignedUrl: true,
+      contentLength: contentLength,
+      payloadHash: payloadHash,
+      serviceConfiguration: serviceConfiguration,
+    );
+  }
+
+  /// {@macro aws_signature_v4.canonical_request}
+  CanonicalRequest._({
+    required this.request,
+    required this.queryParameters,
+    required this.canonicalQueryParameters,
+    required this.headers,
+    required this.canonicalHeaders,
+    required this.signedHeaders,
+    required this.credentialScope,
+    required this.contentLength,
+    required this.payloadHash,
+    required this.serviceConfiguration,
+    required this.presignedUrl,
+    this.algorithm,
+    Duration? expiresIn,
+  })  : normalizePath = serviceConfiguration.normalizePath,
+        omitSessionTokenFromSigning = serviceConfiguration.omitSessionToken,
+        expiresIn = expiresIn?.inSeconds;
+
+  /// Returns the normalized path with double-encoded path segments.
+  ///
+  /// Uses [url] to normalize the path.
+  static String _canonicalPath(
+    AWSBaseHttpRequest request, {
+    required bool normalizePath,
+  }) {
+    var path = normalizePath ? url.normalize(request.path) : request.path;
+
+    // `normalize` removes leading and trailing slashes which should be preserved.
+    if (normalizePath) {
+      if (request.path.endsWith('/')) {
+        path = path.ensureEndsWith('/');
+      }
+      if (!request.path.startsWith('/')) {
+        path = path.ensureStartsWith('/');
+      }
```

Comment:
This is just what the signer expects. Normalized paths must end with slashes if the input path ends with one. And normalized paths must always start with a slash, so we add one if it wasn't there. We look at `request.path` because it's the unnormalized or pre-processed path,

Reply:
Yeah I think what I am getting at is shouldn't we be able to just *always* do `path = path.ensureStartsWith('/');`? It would just make the logical parsing that much simpler here (for readers, the computer doesn't care).

Reply:
Ohh gotcha. Yeah, updated.